decoder.go

```diff
@@ -0,0 +1,211 @@
+package redis
+
+import (
+	"errors"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+// Decoder represents a type with a custom Decode() function that can
+// take a string and parse it to set a desired value on a struct field,
+// similar to how encoding/json implements UnmarshalJSON().
+type Decoder interface {
+	Decode(string) error
+}
```

Comment:
I suggest to drop this interface for now for simplicity. We can add something simimlar later.